docs/src/tutorials/entry-tutorial.md

```diff
@@ -0,0 +1,1085 @@
+
+## Setup
+
+We assume that you have Apalache installed. If not, check the manual page on
+[Apalache installation][]. The minimal required version is 0.22.0.
+
+## Running example: Binary search
+
+We are not going to explain the idea of binary search in this tutorial. If you
+need more context on this, check the Wikipedia page on the [Binary search
+algorithm][]. Let's jump straight in the Java code that is given in [Nearly All
+Binary Searches and Mergesorts are Broken][]:
+
+```java
+1:     public static int binarySearch(int[] a, int key) {
+2:         int low = 0;
+3:         int high = a.length - 1;
+4:
+5:         while (low <= high) {
+6:             int mid = (low + high) / 2;
+7:             int midVal = a[mid];
+8:
+9:             if (midVal < key)
+10:                 low = mid + 1
+11:             else if (midVal > key)
+12:                 high = mid - 1;
+13:             else
+14:                 return mid; // key found
+15:         }
+16:         return -(low + 1);  // key not found.
+17:     }
+```
+
+Fast forward, as was found by Joshua Bloch, the expression in line 6 may throw
+an out of bounds exception, due to an integer overflow. This is because `low`
+and `high` are signed integers whose maximum value could not grow above `2^31 -
+1`, so the sum `low + high` could wrap into a negative number.
+
+Let's see how TLA+ and Apalache can help us here. A bit of warning: The final
+TLA+ specification will happen to be longer than the 17 lines above. Don't get
+disappointed too fast. There are several reasons for that:
+
+ 1. TLA+ is not tuned towards one particular class of algorithms, e.g.,
+ sequential algorithms.
+
+ 2. Related to the previous point, TLA+ and Apalache are not tuned to C or Java
+ programs. A software model checker such as [CBMC][], [Stainless][], or
+ [Coral][] would probably accept a shorter program, and it would check it
+ faster. However, if you have a sledgehammer like TLA+, you don't have to learn
+ other languages.
+
+ 3. We have to explicitly write the expected properties that are normally
+ written in comments.
+
+ 4. We have to introduce a bit of boilerplate, to make Apalache work.
+
+## Step 0: Introducing a template module
+
+Let's open a new file called `BinSearch0.tla` and type the very minimal module
+definition:
+
+```tla
+{{#include ../../../test/tla/bin-search/BinSearch0.tla:1:8}}
+```
+
+This module is doing almost nothing. However, it contains a few important things:
+
+ - It uses three standard modules: `Integers`, `Sequences`, and `Apalache`.
+ - It declares the predicate `Init`. This predicate describes the initial
+   states of our state machine. Since we have not declared any variables, it
+   defines the only single state.
+ - It declares the predicate `Next`. This predicate describes the transitions
+   of our state machine. Again, there are no variables and `Next == TRUE`,
+   so this transition goes from the initial state to itself.
```

Comment:
I mean, technically this isn't true for pure TLA+, right? It goes to any unconstrained state.

Reply:
Right. But since there are no variables, there should be just one state.